[test] Fix flaky data source tests in DataGrid

lauri865 opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

<!-- You can use `## Changelog` to create a description for this change in the next release. -->

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/mui-x/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

Randomly failing datasource tests have given me so much grief. It seems like there's a timing issue with `useMockServer` and datasource tests. I'm not 100% sure if I got to the bottom of it, but the problem is somewhere in `useMockServer` initialisation (doesn't happen synchronously), so every time `fetchRows` function reference changes, `fetchRowSpy` gets reset, but there could be timing issues still where previous fetches happen after `fetchRowSpy` has been reset, but fetch comes in later.

With delayed mounting, and waiting for mockServer, we now have to deal with StrictMode as well, since they don't get reset anymore in the `useMemo` that also runs twice (at least). Added a monkeypatch for now (a <Reset /> component) to see the actual problematic tests, since I didn't want to meddle with every single test.

Ideally `fetchRowSpy` should only be reset once per test, or within the test itself, not within the component lifecycle, otherwise these tests can be very flaky. I think for testing purposes, it would be better if `mockServer` wasn't a hook.

However, now we also have to deal with StrictMode fetches, since with delayed rendering, on mount it will fetch twice (always has, but previously it was reset in `useMemo` that also runs twice, at least, sometimes more).

Edit: fixed the tests.

romgrk commented on packages/x-data-grid-premium/src/tests/dataSourceAggregation.DataGridPremium.test.tsx:
```diff
@@ -22,9 +22,17 @@ describe('<DataGridPremium /> - Data source aggregation', () => {
+  const fetchRowsSpy = spy();
   let mockServer: ReturnType<typeof useMockServer>;
 
+  // TODO: Resets strictmode calls, need to find a better fix for this, maybe an AbortController?
```
I would also be in favor of removing strict mode.

lauri865 replied:
I don't know if you know this, but `useGridRegisterPipeProcessor` is semi-broken in StrictMode. We can't control whether developers use StrictMode though, so turning it off doesn't just fix it.

1. StrictMode call – invokes `useFirstRender` (it will only be invoked in StrictMode, since the `isFirstRender` ref doesn't restore, and we cannot restore it back
2. StrictMode call – runs `useEffect` in `useGridRegisterPipeProcessor`, that sets `isFirstRender` to false, and returns cleanup, that will remove the processor at the end of StrictMode invocation
3. Normal call – `useFirstRender` has no effect, but processor is cleaned up
4. Processor is re-added only in `useEffect`, so anything that calls it before fails. Which is the reason why I reordered the hooks in `useDatagridPremiumComponents`

I'm not sure what a good fix is, besides following the rules of react, and do it in a simple `useLayoutEffect` to minimise the importance of ordering hooks.

We could prevent returning a cleanup function if `isFirstRender` is true, but then the processor never gets cleaned up without StrictMode, if it only registers once. Which may or may not be fine.

lauri865 replied:
https://github.com/tumugin/use-disposable/blob/604502d69680b0574c67d185cbd2254493997426/src/useIsStrictMode.ts#L35

MBilalShafi commented on packages/x-data-grid-generator/src/hooks/useMockServer.ts:
```diff
@@ -367,5 +368,6 @@ export const useMockServer = (
     loadNewData: () => {
       setIndex((oldIndex) => oldIndex + 1);
     },
+    isReady: Boolean(data?.rows?.length),
```
It could probably be used in the demos too, let's leave it for later though.
Side note: When doing `loadNewData` it might not convey the correct information initially since the previous data would be there but outdated, isn't much of a problem though since updating the `fetchRows` would reset the `dataSource` retriggering a fetch call anyways.

MBilalShafi reviewed (approved):
👍

lauri865 commented on packages/x-data-grid-premium/src/DataGridPremium/useDataGridPremiumComponent.tsx:
```diff
@@ -153,8 +153,8 @@ export const useDataGridPremiumComponent = (
   useGridTreeData(apiRef, props);
-  useGridDataSource(apiRef, props);
   useGridAggregation(apiRef, props);
+  useGridDataSource(apiRef, props);
```
It's fine as long as it comes after `useGridAggregation`, but if we fix the below, then the order won't matter (for this use case at least):
https://github.com/mui/mui-x/pull/16395#discussion_r1934837200

lauri865 commented:
@MBilalShafi, I force pushed over the last change, since I fixed the root cause why the test failed – it was a completely valid failure. Throttling was never cleaned up, which caused race conditions when changing models. Try to avoid using `greaterThan` for fetch calls, that will just allow for mistakes. Using `useGridSelector` for state selectors was also potentially causing issues, so I refactored those.

Your commit, if you want to reincorporate any of those changes:
https://github.com/mui/mui-x/pull/16395/commits/bc0dbca0e47dbb6c917bff4e45ff64a0a714fc8d

MBilalShafi commented:
> Try to avoid using greaterThan for fetch calls, that will just allow for mistakes.

Agreed, I also started looking into the root cause, you got it first.
Much appreciated 👍

MBilalShafi reviewed (approved):
LGTM, let's see if @arminmeh has to add something

arminmeh commented:
Looks good. Thanks for the improvements
Maybe `Reset` component could be extracted, but it can also be done later

arminmeh commented:
Cherry-pick failed, so I have added the changes manually to https://github.com/mui/mui-x/pull/16382
